### PR TITLE
EN-5825: unit tests for end of epoch rewards

### PR DESCRIPTION
**Pull request by bogdan-rosianu**

### CODE REVIEW

File: data/state/peerAccount_test.go [hunk 1]
```diff
@@ -563,3 +563,93 @@ func TestPeerAccount_DecreaseValidatorSuccessRateWithJournal(t *testing.T) {
...
+			saveAccountCalled = true
+			return nil
+		},
+	}
+
+	acc, _ := state.NewPeerAccount(&mock.AddressMock{}, tracker)
+	err := acc.IncreaseNumSelectedInSuccessBlocks()
+
+	assert.Nil(t, err)
+	assert.True(t, journalizeCalled)
```
Comment (sasurobert) [inline data/state/peerAccount_test.go]: add check whether the value is the one we want

File: data/state/peerAccount_test.go [hunk 2]
```diff
@@ -563,3 +563,93 @@ func TestPeerAccount_DecreaseValidatorSuccessRateWithJournal(t *testing.T) {
...
+			return nil
+		},
+	}
+
+	acc, _ := state.NewPeerAccount(&mock.AddressMock{}, tracker)
+	err := acc.AddToAccumulatedFees(big.NewInt(37))
+
+	assert.Nil(t, err)
+	assert.True(t, journalizeCalled)
+	assert.True(t, saveAccountCalled)
```
Comment (sasurobert) [inline data/state/peerAccount_test.go]: add check whether the value is the one we want

File: data/state/peerAccount_test.go [hunk 3]
```diff
@@ -563,3 +563,93 @@ func TestPeerAccount_DecreaseValidatorSuccessRateWithJournal(t *testing.T) {
...
+			return nil
+		},
+	}
+
+	acc, _ := state.NewPeerAccount(&mock.AddressMock{}, tracker)
+	err := acc.SetTempRatingWithJournal(100)
+
+	assert.Nil(t, err)
+	assert.True(t, journalizeCalled)
+	assert.True(t, saveAccountCalled)
```
Comment (sasurobert) [inline data/state/peerAccount_test.go]: add check whether the value is the one we want

File: data/state/peerAccount_test.go [hunk 4]
```diff
@@ -563,3 +563,93 @@ func TestPeerAccount_DecreaseValidatorSuccessRateWithJournal(t *testing.T) {
...
+			return nil
+		},
+	}
+
+	acc, _ := state.NewPeerAccount(&mock.AddressMock{}, tracker)
+	err := acc.ResetAtNewEpoch()
+
+	assert.Nil(t, err)
+	assert.Equal(t, 5, numTimesJournalizeWasCalled)
+	assert.Equal(t, 5, numTimesSaveAccountWasCalled)
```
Comment (sasurobert) [inline data/state/peerAccount_test.go]: add check whether the value is the one we want

File: epochStart/mock/nodesCoordinatorMock.go [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+	randomness []byte,
+	round uint64,
+	shardId uint32,
+	epoch uint32,
+) ([]string, error) {
+	if ncm.GetValidatorsPublicKeysCalled != nil {
+		return ncm.GetValidatorsRewardsAddressesCalled(randomness, round, shardId, epoch)
+	}
+
+	validators, err := ncm.ComputeConsensusGroup(randomness, round, shardId, epoch)
```
Comment (sasurobert) [inline epochStart/mock/nodesCoordinatorMock.go]: no need for implementation code here.

File: epochStart/mock/nodesCoordinatorMock.go [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+	randomness []byte,
+	round uint64,
+	shardId uint32,
+	epoch uint32,
+) ([]string, error) {
+	if ncm.GetValidatorsPublicKeysCalled != nil {
+		return ncm.GetValidatorsPublicKeysCalled(randomness, round, shardId, epoch)
+	}
+
+	validators, err := ncm.ComputeConsensusGroup(randomness, round, shardId, epoch)
```
Comment (sasurobert) [inline epochStart/mock/nodesCoordinatorMock.go]: no need for implementation code here.

File: epochStart/mock/nodesCoordinatorMock.go [hunk 7]
```diff
@@ -0,0 +1,155 @@
+package mock
+
+import (
+	"github.com/ElrondNetwork/elrond-go/sharding"
+)
+
+// NodesCoordinatorMock -
+type NodesCoordinatorMock struct {
```
Comment (sasurobert) [inline epochStart/mock/nodesCoordinatorMock.go]: this is a stub. may implement only the relevant functions

File: data/state/peerAccount_test.go [hunk 8]
```diff
@@ -563,3 +563,93 @@ func TestPeerAccount_DecreaseValidatorSuccessRateWithJournal(t *testing.T) {
 	assert.Equal(t, 1, journalizeCalled)
 	assert.Equal(t, 1, saveAccountCalled)
 }
+
+func TestPeerAccount_IncreaseNumSelectedInSuccessBlocks(t *testing.T) {
```
Comment (sasurobert) [inline data/state/peerAccount_test.go]: add test to the 4 new journalizing elements and revert on them. 2 in accounts, 2 in peer accounts

File: data/state/peerAccount_test.go (hunk 1 above)
Comment (bogdan-rosianu) [reply data/state/peerAccount_test.go]: done

File: data/state/peerAccount_test.go (hunk 2 above)
Comment (bogdan-rosianu) [reply data/state/peerAccount_test.go]: done

File: data/state/peerAccount_test.go (hunk 3 above)
Comment (bogdan-rosianu) [reply data/state/peerAccount_test.go]: done

File: data/state/peerAccount_test.go (hunk 4 above)
Comment (bogdan-rosianu) [reply data/state/peerAccount_test.go]: done

File: epochStart/mock/nodesCoordinatorMock.go (hunk 5 above)
Comment (bogdan-rosianu) [reply epochStart/mock/nodesCoordinatorMock.go]: done

File: epochStart/mock/nodesCoordinatorMock.go (hunk 6 above)
Comment (bogdan-rosianu) [reply epochStart/mock/nodesCoordinatorMock.go]: done

File: epochStart/mock/nodesCoordinatorMock.go (hunk 7 above)
Comment (bogdan-rosianu) [reply epochStart/mock/nodesCoordinatorMock.go]: done

File: data/state/peerAccount_test.go (hunk 8 above)
Comment (bogdan-rosianu) [reply data/state/peerAccount_test.go]: done

File: epochStart/metachain/rewards.go [hunk 9]
```diff
@@ -213,9 +213,9 @@ func (r *rewardsCreator) VerifyRewardsMiniBlocks(metaBlock *block.MetaBlock, val
 
 		numReceivedRewardsMBs++
 		createdMiniBlock := createdMiniBlocks[miniBlockHdr.ReceiverShardID]
-		createdMBHash, err := core.CalculateHash(r.marshalizer, r.hasher, createdMiniBlock)
-		if err != nil {
-			return err
+		createdMBHash, errHashCalculate := core.CalculateHash(r.marshalizer, r.hasher, createdMiniBlock)
```
[the remark this answers is not in the document]
Comment (bogdan-rosianu) [reply epochStart/metachain/rewards.go]: because of ineffassign warning of golangci

File: epochStart/metachain/economics.go [hunk 10]
```diff
@@ -222,19 +236,19 @@ func (e *economics) VerifyRewardsPerBlock(
...
 	if computedEconomics.TotalSupply.Cmp(receivedEconomics.TotalSupply) != 0 {
 		return fmt.Errorf("%w total supply computed %d received %d",
-			process.ErrEndOfEpochEconomicsDataDoesNotMatch, computedEconomics.TotalSupply, receivedEconomics.TotalSupply)
+			epochStart.ErrEndOfEpochEconomicsDataDoesNotMatch, computedEconomics.TotalSupply, receivedEconomics.TotalSupply)
 	}
 	if computedEconomics.RewardsPerBlockPerNode.Cmp(receivedEconomics.RewardsPerBlockPerNode) != 0 {
 		return fmt.Errorf("%wrewards per block per node computed %d received %d",
-			process.ErrEndOfEpochEconomicsDataDoesNotMatch, computedEconomics.RewardsPerBlockPerNode, receivedEconomics.RewardsPerBlockPerNode)
+			epochStart.ErrEndOfEpochEconomicsDataDoesNotMatch, computedEconomics.RewardsPerBlockPerNode, receivedEconomics.RewardsPerBlockPerNode)
 	}
```
Comment (sasurobert) [inline epochStart/metachain/economics.go]: add check that previous epoch start metablock hash is the same.
Also add display economics.

File: epochStart/metachain/economics.go [hunk 11]
```diff
@@ -244,3 +256,18 @@ func (e *economics) VerifyRewardsPerBlock(
 func (e *economics) IsInterfaceNil() bool {
 	return e == nil
 }
+
+func logEconomicsDifferences(computed *block.Economics, received *block.Economics) {
+	log.Debug("VerifyRewardsPerBlock error",
```
Comment (sasurobert) [inline epochStart/metachain/economics.go]: log.Warn and make computed and received in separate lines. it would be easier to read.

Comment (bogdan-rosianu) [reply epochStart/metachain/economics.go]: done

File: epochStart/metachain/economics.go (hunk 10 above)
Comment (bogdan-rosianu) [reply epochStart/metachain/economics.go]: done

**Pull request merged**